src/Microsoft.Health.Dicom.Core/Features/Workitem/WorkitemDatasetValidator.ChangeState.cs

```diff
@@ -0,0 +1,104 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Globalization;
+using EnsureThat;
+using FellowOakDicom;
+using Microsoft.Health.Dicom.Core.Extensions;
+using Microsoft.Health.Dicom.Core.Features.Store;
+using Microsoft.Health.Dicom.Core.Features.Workitem.Model;
+using Microsoft.Health.Dicom.Core.Models;
+
+namespace Microsoft.Health.Dicom.Core.Features.Workitem;
+
+public sealed class ChangeWorkitemStateDatasetValidator : WorkitemDatasetValidator
+{
+    // The legal values correspond to the requested state transition. They are: "IN PROGRESS", "COMPLETED", or "CANCELLED".
+    private static readonly IReadOnlySet<ProcedureStepState> AllowedTargetStatesForWorkitemChangeState =
+        new HashSet<ProcedureStepState> { ProcedureStepState.InProgress, ProcedureStepState.Canceled, ProcedureStepState.Completed };
+
+    protected override void OnValidate(DicomDataset dataset)
+    {
+        // Check for missing Transaction UID
+        dataset.ValidateRequirement(DicomTag.TransactionUID, RequirementCode.OneOne);
+
+        // Check for missing Procedure Step State
+        dataset.ValidateRequirement(DicomTag.ProcedureStepState, RequirementCode.OneOne);
+
+        // Check for allowed procedure step state value
+        var targetProcedureStepStateStringValue = dataset.GetString(DicomTag.ProcedureStepState);
+        var targetProcedureStepState = ProcedureStepStateExtensions.GetProcedureStepState(targetProcedureStepStateStringValue);
+        if (!AllowedTargetStatesForWorkitemChangeState.Contains(targetProcedureStepState))
```

Comment:
So if my current state is SCHEDULED and the passed dataset state COMPLETED, how will you throw error. Or if the current state is COMPLETED this will allow to change it to SCHEDULED

Reply:
We need to use the transition table